### ISSUE TITLE
Consolidating filters

### ISSUE THREAD

**Issue by kaathewisegit** [nushell/nushell#11218]
related to
- #10215 

### Selectors

Currently there are the following selectors for the triad (list, record, table).

- `drop` only works on lists and tables, only with indexes (one argument).
- `drop column` only works on tables, only with indexes of columns (one argument).
- `drop nth` only works on lists, only with indexes (rest arguments).
- `get` works with the triad via cell paths, but not with binaries (rest arguments).
- `range` only works on lists, only with ranges (one argument).
- `reject` only works on records and tables, only cell paths (or list of cell paths) (rest arguments).
- `select` same as reject.
- `skip` works on list and tables, one argument.
- `take` works on the triad and binaries, one index argument.
- `first` fetches a single element or several of them.
- `last` is the same as `first`, but for the elements from the end.
- `every` for lists.

### Proposal

My proposal is to merge all of the above into three commands, which work on lists, tables, records, and binaries (by treating binaries like a list of integers).  It is a `select`/`reject` reject pair, which preserves the structure. and `get`, for fetching items while sometimes discarding the structure. 

- `propose select`: select columns and/or rows.  Takes in rest arguments:

   - A single index.  It selects items from lists, rows from tables, and doesn't work on record.  The modulo of the index is taken, so `-1` is interpreted as the last element.
   - A range.  Same as index, but it interprets a list (as in `range`).
   - A cell-path.  Similarly to current `select`, it only accepts top-level names (`ls | select name.0` throws an error).

   The command always returns the same type.  A table is transformed to a table, record to record, list to list, and binary to binary.

- `propose reject`: same as before, but remove the passed rows/columns.

- `propose get` select items.  It accepts all the same arguments `propose select` does.  But it only fetches the values:
   - A single index gets the item from a list or record, int from a binary, or a record from a table.
   - Range works like several single indexes.
   - A cell-path works as usual.

### Rationale

Simplification.  Currently there are nine commands with partially overlapping functionality, and every single one of them has different semantics: one argument vs. ...rest, ranges, types of data they cover, inputting lists of arguments (which `select/reject` allow for, but no other command does).  I think having only two commands and deprecating all the rest would make selecting data easier to lean and do.

**Comment by fdncred** [comment]
There was another issue, pr, or discussion about this recently. One point in that item was the fact that `get` and `select` are different and I don't want to lose them. `get` gets things out of its nushell data structure, `select` leaves things in its nushell data structure.

I'm willing to entertain some of this though because some commands may be better combined with others, but we don't want to have a command with 20 parameters, and we want to keep commands composable. So, sometimes there are commands with too many parameters and those may need to be extracted out to more commands. Also, sometimes a command changes things like `transpose` writing tables or records depending on parameters and those may need to be separated out into multiple commands.

So, for me, it's not about having a bunch of similar commands. I'm ok with having similar commands as long as you know when to use each command.

I'm definitely not down with combining all these commands into two commands.

Also, the `first` and `last` command need to be on your list.

**Comment by KAAtheWiseGit** [comment]
I am conflicted about `first`/`last`.  First is trivial to replace with `select`, but `last` is less obvious.  Would removing it in favor of `select -1` make sense?  Or is the familiarity worth it?  Although, if some kind of merge occurs and `first`/`last` stay, I think they should be reduced to a simple single element selector (without parameters).

**Comment by KAAtheWiseGit** [comment]
And as for get, I see what You mean!  I missed that difference in behavior.  I agree this means that `get` must be kept.

But looking a bit closer, I it would make sense to restrict `get` to one argument.  Because currently `ls | get 0 1` will return a table, which I don't think makes much sense, given this `select` description:

> This differs from `get` in that, rather than accessing the given value in the data structure, it removes all non-selected values from the structure. Hence, using `select` on a table will produce a table, a list will produce a list, and a record will produce a record.

**Comment by fdncred** [comment]
> Would removing it in favor of select -1 make sense? Or is the familiarity worth it?

I'd actually be ok adding `select -1` type of behavior to `select` but then also making `last` just call `select -1` as a helpful shortcut.

>  would make sense to restrict get to one argument. 

I think we should do this consistently across commands. I like the idea of being able to cherry-pick which items you want like `ls | get 1 4 9` but we could make that a list like `let pick = [1 4 9];ls | get $pick`, but also should have the ability with `get` and `select` to do ranges like `ls | range 4..7`.

One thing we also have to keep in mind is dealing with binary data. Nushell is ok working with binary data but any of these filters should have the ability to work with `binary` data. One thing that I find super helpful is these type of "sniffing".
```nushell
❯ open -r Cargo.toml | into binary | take 10 
Length: 10 (0xa) bytes | printable whitespace ascii_other non_ascii
00000000:   5b 70 61 63  6b 61 67 65  5d 0a                      [package]_
❯ open -r Cargo.toml | into binary | first 10 
Length: 10 (0xa) bytes | printable whitespace ascii_other non_ascii
00000000:   5b 70 61 63  6b 61 67 65  5d 0a                      [package]_
```

**Comment by KAAtheWiseGit** [comment]
@fdncred I agree about binaries!

And about `get`, I'm still a bit confused.  What would the difference between `ls | select <range>` and `ls | get <range>` be?

**Comment by fdncred** [comment]
> What would the difference between ls | select <range> and ls | get <range> be?

As it relates to `ls`, there's probably nothing different if you're selecting/getting by row index because `ls` just reports a list of records and when returning multiple records, you typically get a list of records. e.g. `ls | get 0 1 2` should look the same as `ls | select 0 1 2`. However, if there is nested data, things might look different. I'm not sure.

You can see the difference with this test, but this is only with single items selecting/getting by row index.
```
ls | get 0 
╭──────────┬────────────────────╮
│ name     │ CODE_OF_CONDUCT.md │
│ type     │ file               │
│ size     │ 3.4 KB             │
│ modified │ 8 months ago       │
╰──────────┴────────────────────╯
```
```
ls | select 0
╭─#─┬────────name────────┬─type─┬──size──┬───modified───╮
│ 0 │ CODE_OF_CONDUCT.md │ file │ 3.4 KB │ 8 months ago │
╰───┴────────────────────┴──────┴────────┴──────────────╯
```
However, if you're `selecting`/`getting` by column, it's different.
```
ls | get name size | table -e 
╭───┬──────────────────────────────╮
│ 0 │ ╭────┬─────────────────────╮ │
│   │ │  0 │ CODE_OF_CONDUCT.md  │ │
│   │ │  1 │ CONTRIBUTING.md     │ │
│   │ │  2 │ Cargo.lock          │ │
│   │ │  3 │ Cargo.toml          │ │
│   │ │  4 │ Cross.toml          │ │
│   │ │  5 │ LICENSE             │ │
│   │ │  6 │ README.md           │ │
│   │ │  7 │ assets              │ │
│   │ │  8 │ benches             │ │
│   │ │  9 │ crates              │ │
│   │ │ 10 │ devdocs             │ │
│   │ │ 11 │ docker              │ │
│   │ │ 12 │ rust-toolchain.toml │ │
│   │ │ 13 │ scripts             │ │
│   │ │ 14 │ src                 │ │
│   │ │ 15 │ target              │ │
│   │ │ 16 │ tests               │ │
│   │ │ 17 │ toolkit.nu          │ │
│   │ │ 18 │ wix                 │ │
│   │ ╰────┴─────────────────────╯ │
│ 1 │ ╭────┬──────────╮            │
│   │ │  0 │   3.4 KB │            │
│   │ │  1 │  11.4 KB │            │
│   │ │  2 │ 151.2 KB │            │
│   │ │  3 │   6.0 KB │            │
│   │ │  4 │    666 B │            │
│   │ │  5 │   1.1 KB │            │
│   │ │  6 │  12.2 KB │            │
│   │ │  7 │    160 B │            │
│   │ │  8 │    128 B │            │
│   │ │  9 │   1.1 KB │            │
│   │ │ 10 │    192 B │            │
│   │ │ 11 │     96 B │            │
│   │ │ 12 │   1.1 KB │            │
│   │ │ 13 │    448 B │            │
│   │ │ 14 │    448 B │            │
│   │ │ 15 │    224 B │            │
│   │ │ 16 │    512 B │            │
│   │ │ 17 │  15.7 KB │            │
│   │ │ 18 │    160 B │            │
│   │ ╰────┴──────────╯            │
╰───┴──────────────────────────────╯
```
```
❯ ls | select name size
╭─#──┬────────name─────────┬───size───╮
│ 0  │ CODE_OF_CONDUCT.md  │   3.4 KB │
│ 1  │ CONTRIBUTING.md     │  11.4 KB │
│ 2  │ Cargo.lock          │ 151.2 KB │
│ 3  │ Cargo.toml          │   6.0 KB │
│ 4  │ Cross.toml          │    666 B │
│ 5  │ LICENSE             │   1.1 KB │
│ 6  │ README.md           │  12.2 KB │
│ 7  │ assets              │    160 B │
│ 8  │ benches             │    128 B │
│ 9  │ crates              │   1.1 KB │
│ 10 │ devdocs             │    192 B │
│ 11 │ docker              │     96 B │
│ 12 │ rust-toolchain.toml │   1.1 KB │
│ 13 │ scripts             │    448 B │
│ 14 │ src                 │    448 B │
│ 15 │ target              │    224 B │
│ 16 │ tests               │    512 B │
│ 17 │ toolkit.nu          │  15.7 KB │
│ 18 │ wix                 │    160 B │
╰─#──┴────────name─────────┴───size───╯
```
`get` pulls the data out and `select` tries to maintain the shape of the data.

**Comment by KAAtheWiseGit** [comment]
@fdncred got it!  The `get name size` example did it for me, I can now comprehend the difference.  I'll edit the proposal.

**Comment by amtoine** [comment]
@fdncred 
maybe https://github.com/nushell/nushell/issues/10215 is the other issue? :innocent:

**Comment by KAAtheWiseGit** [comment]
I resumed the work on the PR today, and noticed that the current `select` and `reject` implementations collect streams.  So, commands such as `generate null {{out: {a: 1, b: 2, c: 3}, next: null}} | select a | first 10` hang forever[^oom].

[^oom]: I accidentally killed my session with OOM while I was writing this comment, because I forgot to interrupt this operation.

**Comment by nome** [comment]
I like the direction of the proposal. Yesterday I independently came up with a more conservative variant of it (#13779); but I think if this one here gets implemented, it would make nushell even more fun and easier to learn.

One addition: For lists and tables, there's also the `every` selector (and `every --skip`). How does it fit into the proposal?

* `every`/`every --skip` could be renamed to `select every` and `reject every`, in order to keep all selectors in the select/reject/get triad
* alternatively, it could be expressed using the extended range syntax `start..stride..end` - e.g. `every 2` would become `select 0..2..-1`

**Comment by NotTheDr01ds** [comment]
I was also taking a fresh look at this based on #13779.  @KAAtheWiseGit - A few questions:

* Re:*" it only accepts top-level names (ls | select name.0 throws an error)."* And why is this exactly?  Why does (or should) `select name 0` work, but `select name.0` throw an error?  Why should `0.name` work with `get` but not `select`?

  Why should they not both take the same cell-path constructs?

* `select` single index - re:*"doesn't work on record"* - Why not?  It does today.

  ```nu
  { a: 5, b: 10 } | select a
  # =>
  ╭───┬───╮
  │ a │ 5 │
  ╰───┴───╯

  ls | get 0 | get name
  # =>
  CNAME
  ```

**Comment by NotTheDr01ds** [comment]
Also, I think the spec needs to be more clear about what happens with **multiple rest arguments**, even if it just confirms that they should work the same way they do today, which is:

**Multiple** Rest Arguments Behavior:

  * `select`
    * Table -> Table
      * Multiple column names result in a new table with just those columns (e.g., `'ls | select name size`)
      * Multiple row numbers result in a new table with just those rows
      * Row numbers and column names results in the cumulative intersection (looking for a better word) of those columns and rows. E.g.:
        * `ls | select name size 0 -1` the first and last files with the `name` and `size` columns only)
        * `ls | select name 0..4` (using your new range argument, selects the first 5 filenames -- The same as `ls | select name | range 0..4` today)

    * List -> List
      * (Column names are N/A - Attempt to `select <string>` is an error)
      * Multiple row numbers results in a new list with just those rows (`[ zero one two three four five ] | select 0 4`

    * Record -> Record
      * Multiple key-names results in a new record with just those keys (`{ a: foo, b: bar, c: baz } | select a c`)
      * (Row numbers are N/A - Attempt to `select <number>` is an error)

  * `get` - Multiple rest values always result in a `list` of the results
    * Table -> List
      * Multiple column names results in a new **list of lists** of the values in those columns (e.g., `ls | get name size`).  While I'm not sure how useful this is (I've never used it, but I'll be on the lookout for a use-case), it's the way it works today.
      * Multiple row numbers results in a new **list of records** (i.e., a table).  Note that the result of a `get` and `select` are identical in this case.
      * Row numbers and column names results in a new **list of lists** of the values of each `get` taken separately.  E.g., `ls | get name 0 4` (again, is this useful? Maybe?)

  * List -> List
    * (Column names are N/A, as above)
    * Multiple row numbers results in a new **list of values** where each value is the cumulative result of each `get <row_num>`.  E.g., `[ zero one two three four five ] | get 0 4` - As above, note that this is the same result that `select` would return.

  * Record -> List
    * Multiple key names results in a new **list of values** where each value is the cumulative result of each `get <key_name>`.  E.g., `{ a: foo, b: bar, c: baz } | get a c`
    * (Row numbers are N/A, as above)

**Comment by NotTheDr01ds** [comment]
@nome 

> alternatively, it could be expressed using the extended range syntax

Right - I think, given @KAAtheWiseGit's proposal supports ranges as a selector, that this would be the solution

**Comment by KAAtheWiseGit** [comment]
Yeah, `every 2` would be `select 0..2..` and `every 2 --skip` would be `reject 0..2..`

**Comment by KAAtheWiseGit** [comment]
Yep, added

**Comment by KAAtheWiseGit** [comment]
About the questions:

- Disabling `select` for non-top level things.

  The issue here is that `select` does ad-hoc flattening.  At the time
  of filing this issue it would concatenate the path names with an
  underscore.  It uses a dot now, but I still think that this is an
  inelegant way to do this.

  Thinking about it now `select` should probably support cell paths.
  But I think that it should keep the structure, i.e.

  ```Nushell
  ~> [[column]; [{nested: [0 1 2]}] [{nested: [0 1 2]}]] | propose select nested.1
  ───┬────────────────────
   # │       column
  ───┼────────────────────
   0 │ ────────┬─────────
     │         │ ───┬───
     │  nested │  0 │ 1
     │         │ ───┴───
     │ ────────┴─────────
   1 │ ────────┬─────────
     │         │ ───┬───
     │  nested │  0 │ 1
     │         │ ───┴───
     │ ────────┴─────────
  ───┴────────────────────
  ```

- Same idea with `select` on records.  I think I didn't know it worked
  with records at the time.  I now think it should work, but that it
  also should preserve the structure:

  ```Nushell
  ~> {a: {b: c} d: e} | propose select a.b
  ───┬─────────
     │ ───┬───
   a │  b │ c
     │ ───┴───
  ───┴─────────
  ```

- `propose reject` --- it's poorly written, I wanted to refer to
  `propose select`.  So, "same as `propose select`, but ...".

TLDR, I changed my mind since then, `propose select` does need to work
with cell paths, but I think it should preserve the structure.  The
current implementation of already `reject` does:

```Nushel
~> {a: {b: c} d: e} | reject a.b
───┬───────────────────
 a │ {record 0 fields}
 d │ e
───┴───────────────────
```

**Comment by KAAtheWiseGit** [comment]
As for the spec, I think it's exactly how it should work.  I'd be up for dropping it as a Markdown file somewhere in the tree if I manage to finish the rework PR

I find multi-argument `get` a bit quirky too, but I shouldn't be too hard to implement (by simply calling the single-argument `get` function in a loop), so I'm fine with it

**Comment by weirdan** [comment]
> but I think it should preserve the structure.

I think I've got to voice a disagreement here. `select` is a nice projection function (even though it does not support column renaming, as its SQL counterpart does), and forcing it to keep the structure makes it way less useful.

**Comment by KAAtheWiseGit** [comment]
I personally see the structure preservation as a core "thing" of the `select` command.  This way it's symmetrical to `reject`.

I think that in this case there should be a separate projection function, with documented and configurable behavior, rather than simple dot joins.

**Comment by NotTheDr01ds** [comment]
> is a nice projection function

@weirdan  Can you give an example to demonstrate?  In general, I thought `select` already (and should continue to) "keep the structure" - `table -> table`, `list -> list`, and `record -> record`

**Comment by KAAtheWiseGit** [comment]
@NotTheDr01ds `select` will flatten nested cell paths

```Nushell
~> [[column]; [{nested: [0 1 2]}] [{nested: [0 1 2]}]] | select column.nested.1
───┬─────────────────
 # │ column.nested.1 
───┼─────────────────
 0 │               1 
 1 │               1 
───┴─────────────────
```

**Comment by NotTheDr01ds** [comment]
> @NotTheDr01ds `select` will flatten nested cell paths

Okay - I get it now.  Had no idea that even worked since `select column.1` by itself errors out.  That's a weird bug.

Given that, in your example:

> Thinking about it now select should probably support cell paths. But I think that it should keep the structure, i.e.
>
> ```nu
> [[column]; [{nested: [0 1 2]}] [{nested: [0 1 2]}]] | propose select nested.1
> ```

Do you mean `propose select column.nested.1`?

**Comment by KAAtheWiseGit** [comment]
Ah, yes, forgot to add column.  Lemme fix that

**Comment by NotTheDr01ds** [comment]
>  I'd be up for dropping it as a Markdown file somewhere in the tree if I manage to finish the rework PR

I'm kind of writing it up that way in preparation for eventually updating the docs (when you finish it, yes).

I'll admit I started out skeptical of this, since I find the "convenience" versions like `first` and `drop` so much more readable. But perhaps they can be recreated as custom commands and moved to `std-lib` (also once those performance issues get addressed).

**Comment by nome** [comment]
> Yeah, `every 2` would be `select 0..2..` and `every 2 --skip` would be `reject 0..2..`

Come to think of it, at least one programming language that I currently don't remember (or was it a computer algebra system?)  supports the range syntax `first,second..last`, where `stride = second - first` is computed implicitly. So, for example, `1,3..7` would expand to `[1,3,5,7]`. This might be an alternative to consider. Benefit: More intuitive, since it's closer to how one might write down an ellipsised range of numbers on paper. Drawback: Might introduce parsing ambiguity?

**Comment by KAAtheWiseGit** [comment]
@nome yeah, there was a medium-sized discussion about it in the past.  I was skeptical of the stride thing initially, but the example with `every` showed me that there are applications for it.  Writing `select 0..2..` is much more convenient than `select 0..inf..2`

**Comment by Iron-E** [comment]
Was this implemented, or is it no longer planned?

**Comment by kaathewisegit** [comment]
I think I was the only one championing this, but I no longer think that such a big redesign is a good idea now that Nushell is mostly mature
